# Génération des embeddings sémantiques

Cette étape consiste à transformer le référentiel de compétences construit dans la couche Gold en vecteurs sémantiques.

Ces vecteurs permettront :

- de comparer un profil utilisateur à notre référentiel
- d’évaluer la similarité entre compétences et expérience utilisateur
- d’alimenter le moteur de recommandation métier

Le modèle utilisé est Sentence-BERT (SBERT).

In [1]:
import pandas as pd
import numpy as np
from sentence_transformers import SentenceTransformer

# Chargement du référentiel de compétences

In [2]:
competencies_df = pd.read_csv(
    "../data/gold/competencies.csv"
)

competencies_df.head()

,competency_id,competency,block,semantic_description
0,C001,AWS,Cloud & DevOps,Expérience dans le déploiement d'infrastructur...
1,C002,Agile,Software Engineering Practices,Expérience du travail en équipe agile et de la...
2,C003,CI/CD,Cloud & DevOps,Capacité à automatiser les processus de compil...
3,C004,Docker,Cloud & DevOps,Capacité à conteneuriser des applications et à...
4,C005,Git,Cloud & DevOps,"Capacité à versionner, gérer et collaborer sur..."


# Chargement du modèle Sentence-BERT

In [3]:
model = SentenceTransformer(
    "paraphrase-multilingual-MiniLM-L12-v2",
    device="cpu"
)

modules.json:   0%|          | 0.00/229 [00:00<?, ?B/s]

e:\ECOLE\AISCA_PROJECT\venv\Lib\site-packages\huggingface_hub\file_download.py:137: UserWarning: `huggingface_hub` cache-system uses symlinks by default to efficiently store duplicated files but your machine does not support them in C:\Users\arnau\.cache\huggingface\hub\models--sentence-transformers--paraphrase-multilingual-MiniLM-L12-v2. Caching files will still work but in a degraded version that might require more space on your disk. This warning can be disabled by setting the `HF_HUB_DISABLE_SYMLINKS_WARNING` environment variable. For more details, see https://huggingface.co/docs/huggingface_hub/how-to-cache#limitations.
To support symlinks on Windows, you either need to activate Developer Mode or to run Python as an administrator. In order to activate developer mode, see this article: https://docs.microsoft.com/en-us/windows/apps/get-started/enable-your-device-for-development
  warnings.warn(message)


config_sentence_transformers.json:   0%|          | 0.00/122 [00:00<?, ?B/s]

README.md:   0%|          | 0.00/3.89k [00:00<?, ?B/s]

sentence_bert_config.json:   0%|          | 0.00/53.0 [00:00<?, ?B/s]

config.json:   0%|          | 0.00/645 [00:00<?, ?B/s]

model.safetensors:   0%|          | 0.00/471M [00:00<?, ?B/s]

Loading weights:   0%|          | 0/199 [00:00<?, ?it/s]

tokenizer_config.json:   0%|          | 0.00/526 [00:00<?, ?B/s]

tokenizer.json:   0%|          | 0.00/9.08M [00:00<?, ?B/s]

special_tokens_map.json:   0%|          | 0.00/239 [00:00<?, ?B/s]

config.json:   0%|          | 0.00/190 [00:00<?, ?B/s]

# Extraction des descriptions sémantiques

In [4]:
semantic_texts = competencies_df[
    "semantic_description"
].tolist()

print(semantic_texts)

["Expérience dans le déploiement d'infrastructures cloud et d'applications avec Amazon Web Services.", 'Expérience du travail en équipe agile et de la gestion de projet par itérations.', 'Capacité à automatiser les processus de compilation, de test et de déploiement logiciel avec des pipelines CI/CD.', 'Capacité à conteneuriser des applications et à gérer des environnements reproductibles avec Docker.', 'Capacité à versionner, gérer et collaborer sur du code source avec Git.', 'Capacité à développer des services backend et des applications avec le langage Go.', "Capacité à développer des applications backend ou d'entreprise avec Java.", 'Capacité à développer des applications web interactives avec JavaScript.', 'Capacité à orchestrer et déployer des applications conteneurisées avec Kubernetes.', 'Capacité à entraîner, évaluer et déployer des modèles de machine learning pour des tâches prédictives.', 'Capacité à stocker, interroger et gérer des données NoSQL avec MongoDB.', 'Capacité à 

# Génération des vecteurs sémantiques

In [5]:
competency_embeddings = model.encode(
    semantic_texts
)

In [6]:
print(competency_embeddings.shape)

(19, 384)


# Sauvegarde des embeddings

In [7]:
np.save(
    "../data/gold/competency_embeddings.npy",
    competency_embeddings
)

In [8]:
loaded_embeddings = np.load(
    "../data/gold/competency_embeddings.npy"
)

print(loaded_embeddings.shape)

(19, 384)


# Création d’un profil utilisateur fictif

In [9]:
user_profile = """

I have experience building machine learning systems.

I deployed cloud applications using AWS.

I worked with Docker containers.

I developed automation scripts in Python.

"""

In [10]:
user_embedding = model.encode(
    user_profile
)

In [11]:
print(user_embedding.shape)

(384,)


# Conclusion

Le référentiel de compétences a été transformé en vecteurs sémantiques.

Chaque compétence dispose désormais d’une représentation vectorielle de 384 dimensions.

Ces embeddings seront utilisés dans l’étape suivante afin de calculer la similarité entre le profil utilisateur et les compétences du référentiel.

La prochaine étape consistera à construire le moteur de similarité sémantique.